pgrx/src/datum/borrow.rs

```diff
@@ -0,0 +1,104 @@
+    /// The "native" passing convention for this type.
+    ///
+    /// - `PassBy::Value` implies [`mem::size_of<T>()`][size_of] <= [`mem::size_of::<Datum>()`][Datum].
+    /// - `PassBy::Ref` means the pointee will occupy at least 1 byte for variable-sized types.
+    ///
+    /// Note that this means a zero-sized type is inappropriate for `BorrowDatum`.
+    const PASS: PassBy;
+
+    /// Cast a pointer to this blob of bytes to a pointer to this type.
+    ///
+    /// This is not a simple `ptr.cast()` because it may be *unsizing*, which may require
+    /// reading varlena headers. For all fixed-size types, `ptr.cast()` should be correct.
+    ///
+    /// # Safety
+    /// - This must be correctly invoked for the pointee type, as it may deref and read one or more
+    ///   bytes in its implementation in order to read the inline metadata and unsize the type.
+    ///
+    /// ## For Implementers
+    /// Reading the **first** byte pointed to is permitted if `T::PASS = PassBy::Ref`, assuming you
+    /// are implementing a varlena type. As the other dynamic length type, CStr also does this.
+    /// This function
+    /// - must NOT mutate the pointee
+    /// - must return a pointer with provenance to the entire allocation
+    ///
+    /// Do not attempt to handle pass-by-value versus pass-by-ref in this fn's body!
+    /// A caller may be in a context where all types are handled by-reference, for instance.
+    unsafe fn point_from(ptr: ptr::NonNull<u8>) -> ptr::NonNull<Self>;
+
+    /// Cast a pointer to aligned varlena headers to this type
+    ///
+    /// This version allows you to assume the pointer is aligned to, and readable for, 4 bytes.
+    /// This optimization is not required. When in doubt, avoid implementing it, and rely on your
+    /// `point_from` implementation alone.
+    ///
+    /// # Safety
+    /// - This must be correctly invoked for the pointee type, as it may deref.
+    /// - This must be 4-byte aligned!
+    unsafe fn point_from_align4(ptr: ptr::NonNull<u32>) -> ptr::NonNull<Self> {
+        debug_assert!(ptr.is_aligned());
+        unsafe { <Self as BorrowDatum>::point_from(ptr.cast()) }
+    }
+
+    /// Optimization for borrowing the referent
+    unsafe fn borrow_unchecked<'dat>(ptr: ptr::NonNull<u8>) -> &'dat Self {
+        unsafe { Self::point_from(ptr).as_ref() }
+    }
+}
+
+macro_rules! borrow_by_value {
+    ($($value_ty:ty),*) => {
+        $(
+            unsafe impl BorrowDatum for $value_ty {
+                const PASS: PassBy = if mem::size_of::<Self>() <= mem::size_of::<Datum>() {
+                    PassBy::Value
+                } else {
+                    PassBy::Ref
+                };
+
+                unsafe fn point_from(ptr: ptr::NonNull<u8>) -> ptr::NonNull<Self> {
+                    ptr.cast()
+                }
+            }
+        )*
+    }
+}
+
+borrow_by_value! {
+    i8, i16, i32, i64, bool, f32, f64, pg_sys::Oid, Date, Time, Timestamp, TimestampWithTimeZone
+}
```

Comment:
This needs more tests that specifically exercise BorrowDatum on these "by-value" cases, and not just its interaction with FlatArray or other varlenas. Varlenae?

Reply:
I couldn't find a good way to test this because it requires either FnCall support or SPI support, and I don't want to extend FromDatum + IntoDatum to cover these cases yet.

Reply:
nevermind